File: cslacs/Templates/EditableChildList.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Data.SqlClient;
using Csla;

namespace Templates
{
  [Serializable()]
  class EditableChildList : 
    BusinessListBase<EditableChildList, EditableChild>
  {
    #region Factory Methods

    internal static EditableChildList NewEditableChildList()
    {
      return new EditableChildList();
    }

    internal static EditableChildList GetEditableChildList(
      SqlDataReader dr)
    {
      return new EditableChildList(dr);
    }

    private EditableChildList()
    {
      MarkAsChild();
    }

    private EditableChildList(SqlDataReader dr)
    {
      MarkAsChild();
      Fetch(dr);
    }

    #endregion

    #region Data Access

    private void Fetch(SqlDataReader dr)
    {
      RaiseListChangedEvents = false;
      while (dr.Read())
      {
        this.Add(EditableChild.GetEditableChild(dr));
      }
      RaiseListChangedEvents = true;
    }

    internal void Update(object parent)
    {
      RaiseListChangedEvents = false;
      foreach (EditableChild item in DeletedList)
        item.DeleteSelf();
      DeletedList.Clear();

      foreach (EditableChild item in this)
        if (item.IsNew)
          item.Insert(parent);
        else
          item.Update(parent);
      RaiseListChangedEvents = true;
    }

    #endregion
  }
}

File: cslavb/Templates/EditableChildList.vb
Imports System.Data.SqlClient

<Serializable()> _
Public Class EditableChildList
  Inherits BusinessListBase(Of EditableChildList, EditableChild)

#Region " Factory Methods "

  Friend Shared Function NewEditableChildList() As EditableChildList
    Return New EditableChildList
  End Function

  Friend Shared Function GetEditableChildList( _
    ByVal dr As SqlDataReader) As EditableChildList

    Return New EditableChildList(dr)
  End Function

  Private Sub New()
    MarkAsChild()
  End Sub

  Private Sub New(ByVal dr As SqlDataReader)
    MarkAsChild()
    Fetch(dr)
  End Sub

#End Region

#Region " Data Access "

  Private Sub Fetch(ByVal dr As SqlDataReader)

    RaiseListChangedEvents = False
    While dr.Read
      Add(EditableChild.GetEditableChild(dr))
    End While
    RaiseListChangedEvents = True

  End Sub

  Friend Sub Update(ByVal parent As Object)

    RaiseListChangedEvents = False
    For Each item As EditableChild In DeletedList
      item.DeleteSelf()
    Next
    DeletedList.Clear()

    For Each item As EditableChild In Me
      If item.IsNew Then
        item.Insert(parent)

      Else
        item.Update(parent)
      End If
    Next
    RaiseListChangedEvents = True

  End Sub

#End Region

End Class
